Rework `untranslatable_diagnostic` lint

nnethercote opened the pull request:
Currently it only checks calls to functions marked with `#[rustc_lint_diagnostics]`. This PR changes it to check calls to any function with an `impl Into<{D,Subd}iagnosticMessage>` parameter. This greatly improves its coverage and doesn't rely on people remembering to add `#[rustc_lint_diagnostics]`. It also lets us add `#[rustc_lint_diagnostics]` to a number of functions that don't have an `impl Into<{D,Subd}iagnosticMessage>`, such as `Diag::span`.

r? @davidtwco

nnethercote committed:
Rewrite the `untranslatable_diagnostic` lint.

Currently it only checks calls to functions marked with
`#[rustc_lint_diagnostics]`. This commit changes it to check calls to
any function with an `impl Into<{D,Subd}iagMessage>` parameter. This
greatly improves its coverage and doesn't rely on people remembering to
add `#[rustc_lint_diagnostics]`.

The commit also adds `#[allow(rustc::untranslatable_diagnostic)`]
attributes to places that need it that are caught by the improved lint.
These places that might be easy to convert to translatable diagnostics.

Finally, it also:
- Expands and corrects some comments.
- Does some minor formatting improvements.
- Adds missing `DecorateLint` cases to
  `tests/ui-fulldeps/internal-lints/diagnostics.rs`.

nnethercote committed:
Change message type in bug functions.

From `impl Into<DiagnosticMessage>` to `impl Into<Cow<'static, str>>`.

Because these functions don't produce user-facing output and we don't
want their strings to be translated.

nnethercote commented:
@davidtwco: this isn't ready to merge, mostly because the lint code itself is still very hacky. But I'd like to hear what you think of the general idea. It's clear that the new version of the lint finds lots of things the old version doesn't.

I also wonder if the `diagnostic_outside_of_impl` lint is even necessary any more. I'm struggling to think of what problematic code it could find that the expanded `untranslated_diagnostic` couldn't find.

davidtwco commented:
I think this is a great idea, I hadn't thought to check for the `Into<DiagnosticMessage>` when writing this initially and it is much less fragile.

`diagnostic_outside_of_impl` was just intended to steer new diagnostics towards using structs instead of the builder API - it's unclear how much we want this, we're going to be discussing that in wg-diagnostics soon, but for now it should probably stay.

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/8015468043/job/21895684067) [(plain)](https://github.com/rust-lang/rust/commit/ccd24fa21927af30679b6613b37cea2e14d3fb2f/checks/21895684067/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Prepare all required actions
Getting action download info
Download action repository 'actions/checkout@v4' (SHA:b4ffde65f46336ab88eb53be808477a3936bae11)
Download action repository 'actions/upload-artifact@v3' (SHA:a8a3f3ad30e3422c9c7b888a15615d19a852ae32)
Complete job name: PR - mingw-check-tidy
git config --global core.autocrlf false
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
---
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_69519d4f-d980-493d-b7db-c67ebb83b8f2
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_PATH=/home/runner/work/_temp/_github_workflow/event.json
GITHUB_GRAPHQL_URL=https://api.github.com/graphql
GITHUB_HEAD_REF=rework-untranslatable_diagnostic-lint
GITHUB_JOB=pr
GITHUB_PATH=/home/runner/work/_temp/_runner_file_commands/add_path_69519d4f-d980-493d-b7db-c67ebb83b8f2
GITHUB_REF=refs/pull/121382/merge
GITHUB_REF_NAME=121382/merge
GITHUB_REF_PROTECTED=false
---
COPY scripts/sccache.sh /scripts/
RUN sh /scripts/sccache.sh

COPY host-x86_64/mingw-check/reuse-requirements.txt /tmp/
RUN pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt \
    && pip3 install virtualenv
COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---

#10 [5/8] COPY host-x86_64/mingw-check/reuse-requirements.txt /tmp/
#10 DONE 0.0s

#11 [6/8] RUN pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt     && pip3 install virtualenv
#11 0.696   Downloading binaryornot-0.4.4-py2.py3-none-any.whl (9.0 kB)
#11 0.714 Collecting boolean-py==4.0
#11 0.721   Downloading boolean.py-4.0-py3-none-any.whl (25 kB)
#11 0.740 Collecting chardet==5.1.0
---
#11 4.020 Building wheels for collected packages: reuse
#11 4.021   Building wheel for reuse (pyproject.toml): started
#11 4.373   Building wheel for reuse (pyproject.toml): finished with status 'done'
#11 4.374   Created wheel for reuse: filename=reuse-1.1.0-cp310-cp310-manylinux_2_35_x86_64.whl size=181117 sha256=f5f58750481f69515c2c0d1d503daf565e2565c370d07fc6aeb95fe3498b4269
#11 4.374   Stored in directory: /tmp/pip-ephem-wheel-cache-_z3l2hyk/wheels/c2/3c/b9/1120c2ab4bd82694f7e6f0537dc5b9a085c13e2c69a8d0c76d
#11 4.377 Installing collected packages: boolean-py, binaryornot, setuptools, reuse, python-debian, markupsafe, license-expression, jinja2, chardet
#11 4.400   Attempting uninstall: setuptools
#11 4.401     Found existing installation: setuptools 59.6.0
#11 4.402     Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
#11 4.402     Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
#11 4.403     Can't uninstall 'setuptools'. No files were found to uninstall.
#11 5.121 Successfully installed binaryornot-0.4.4 boolean-py-4.0 chardet-5.1.0 jinja2-3.1.2 license-expression-30.0.0 markupsafe-2.1.1 python-debian-0.1.49 reuse-1.1.0 setuptools-66.0.0
#11 5.121 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#11 5.666 Collecting virtualenv
#11 5.722   Downloading virtualenv-20.25.1-py3-none-any.whl (3.8 MB)
#11 5.895      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.8/3.8 MB 22.0 MB/s eta 0:00:00
#11 5.948 Collecting platformdirs<5,>=3.9.1
#11 5.956   Downloading platformdirs-4.2.0-py3-none-any.whl (17 kB)
#11 5.977 Collecting distlib<1,>=0.3.7
#11 5.986   Downloading distlib-0.3.8-py2.py3-none-any.whl (468 kB)
#11 6.032 Collecting filelock<4,>=3.12.2
#11 6.040   Downloading filelock-3.13.1-py3-none-any.whl (11 kB)
#11 6.040   Downloading filelock-3.13.1-py3-none-any.whl (11 kB)
#11 6.127 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#11 6.297 Successfully installed distlib-0.3.8 filelock-3.13.1 platformdirs-4.2.0 virtualenv-20.25.1
#11 DONE 6.4s

#12 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#12 DONE 0.0s
---
DirectMap4k:      171968 kB
DirectMap2M:     6119424 kB
DirectMap1G:    12582912 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint
    Finished dev [unoptimized] target(s) in 0.03s
##[endgroup]
downloading https://ci-artifacts.rust-lang.org/rustc-builds-alt/71ff1c62469cc733855a8914976ee9e26f9fe7cc/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/llvm-71ff1c62469cc733855a8914976ee9e26f9fe7cc-true/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm
---
   Compiling tidy v0.1.0 (/checkout/src/tools/tidy)
    Finished release [optimized] target(s) in 29.64s
##[endgroup]
fmt check
##[error]Diff in /checkout/compiler/rustc_lint/src/internal.rs at line 2:
 //! Clippy.
 use crate::lints::{
 use crate::lints::{
-    BadOptAccessDiag, DefaultHashTypesDiag, DiagOutOfImpl, LintPassByHand,
-    NonExistentDocKeyword, QueryInstability, SpanUseEqCtxtDiag, TyQualified, TykindDiag,
-    TykindKind, UntranslatableDiag,
+    BadOptAccessDiag, DefaultHashTypesDiag, DiagOutOfImpl, LintPassByHand, NonExistentDocKeyword,
+    QueryInstability, SpanUseEqCtxtDiag, TyQualified, TykindDiag, TykindKind, UntranslatableDiag,
     UntranslatableDiagnosticTrivial,
 };
 use crate::{EarlyContext, EarlyLintPass, LateContext, LateLintPass, LintContext};
Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/rustfmt/bin/rustfmt" "--config-path" "/checkout" "--edition" "2021" "--unstable-features" "--skip-children" "--check" "/checkout/compiler/rustc_lint/src/foreign_modules.rs" "/checkout/compiler/rustc_lint/src/errors.rs" "/checkout/compiler/rustc_errors/src/json/tests.rs" "/checkout/compiler/rustc_errors/src/lib.rs" "/checkout/compiler/rustc_errors/src/diagnostic_impls.rs" "/checkout/compiler/rustc_errors/src/error.rs" "/checkout/compiler/rustc_errors/src/json.rs" "/checkout/compiler/rustc_lint/src/internal.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
  local time: Fri Feb 23 06:23:06 UTC 2024
  network time: Fri, 23 Feb 2024 06:23:07 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

nnethercote commented:
r? @Nilstrieb, mostly for the `check_expr` code which is the part I'm not sure about and has some `njn` comments indicating things that need polish.

bors commented:
:umbrella: The latest upstream changes (presumably #121741) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

nnethercote committed:
Add missing `#[rustc_lint_diagnostics]` attributes.

Prior to the previous commit, `#[rust_lint_diagnostics]` attributes
could only be used on methods with an `impl Into<{D,Subd}iagMessage>`
parameter. But there are many other nearby diagnostic methods (e.g.
`Diag::span`) that don't take such a parameter and should have the
attribute.

This commit adds the missing attribute to these `Diag` methods. This
requires adding some missing
`#[allow(rustc::diagnostic_outside_of_impl)]` markers at call sites to
these methods.

nnethercote commented:
I rebased.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-16`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/8089344568/job/22105063437) [(plain)](https://github.com/rust-lang/rust/commit/de294affc3c2086d16496c622ebb8eb98fb68177/checks/22105063437/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
GITHUB_ENV=/home/runner/work/_temp/_runner_file_commands/set_env_889de9a7-e4d2-4f44-8eb0-f2e0cd44dbc2
GITHUB_EVENT_NAME=pull_request
GITHUB_EVENT_PATH=/home/runner/work/_temp/_github_workflow/event.json
GITHUB_GRAPHQL_URL=https://api.github.com/graphql
GITHUB_HEAD_REF=rework-untranslatable_diagnostic-lint
GITHUB_JOB=pr
GITHUB_PATH=/home/runner/work/_temp/_runner_file_commands/add_path_889de9a7-e4d2-4f44-8eb0-f2e0cd44dbc2
GITHUB_REF=refs/pull/121382/merge
GITHUB_REF_NAME=121382/merge
GITHUB_REF_PROTECTED=false
---
#12 writing image sha256:9086f70991f336ca4e364e28f0b9963c1b871b4eb3a33c328f7004b17731ef84 done
#12 naming to docker.io/library/rust-ci done
#12 DONE 10.0s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-16]
##[group]Clock drift check
  local time: Thu Feb 29 00:59:53 UTC 2024
  network time: Thu, 29 Feb 2024 00:59:53 GMT
  network time: Thu, 29 Feb 2024 00:59:53 GMT
##[endgroup]
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-16', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'build.optimized-compiler-builtins', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-16/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---
-   --> $DIR/diagnostics.rs:42:9
+ error[E0412]: cannot find type `DiagnosticBuilder` in this scope
+   --> $DIR/diagnostics.rs:83:48
3    |
- LL |         Diag::new(dcx, level, "untranslatable diagnostic")
-    |
- note: the lint level is defined here
-   --> $DIR/diagnostics.rs:6:9
-    |
-    |
- LL | #![deny(rustc::untranslatable_diagnostic)]
-    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+ LL |     fn decorate_lint<'b, >(self, diag: &'b mut DiagnosticBuilder<'a, ()>) {
Build completed unsuccessfully in 0:13:56
12 
- error: diagnostics should be created using translatable messages
-   --> $DIR/diagnostics.rs:63:14
-   --> $DIR/diagnostics.rs:63:14
+ error[E0412]: cannot find type `DiagnosticBuilder` in this scope
+   --> $DIR/diagnostics.rs:96:46
15    |
- LL |         diag.note("untranslatable diagnostic");
-    |              ^^^^
+ LL |     fn decorate_lint<'b>(self, diag: &'b mut DiagnosticBuilder<'a, ()>) {
18 
- error: diagnostics should be created using translatable messages
-   --> $DIR/diagnostics.rs:84:14
-    |
-    |
- LL |         diag.note("untranslatable diagnostic");
-    |              ^^^^
+ error: aborting due to 2 previous errors
24 
- error: diagnostics should only be created in `IntoDiagnostic`/`AddToDiagnostic` impls
-   --> $DIR/diagnostics.rs:106:21
-    |
- LL |     let _diag = dcx.struct_err(crate::fluent_generated::no_crate_example);
-    |
- note: the lint level is defined here
-   --> $DIR/diagnostics.rs:7:9
-    |
-    |
- LL | #![deny(rustc::diagnostic_outside_of_impl)]
-    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
- 
- error: diagnostics should only be created in `IntoDiagnostic`/`AddToDiagnostic` impls
-   --> $DIR/diagnostics.rs:109:21
-    |
- LL |     let _diag = dcx.struct_err("untranslatable diagnostic");
- 
- error: diagnostics should be created using translatable messages
-   --> $DIR/diagnostics.rs:109:21
-    |
-    |
- LL |     let _diag = dcx.struct_err("untranslatable diagnostic");
- 
- error: aborting due to 6 previous errors
- 
+ For more information about this error, try `rustc --explain E0412`.
---
--- stderr -------------------------------
error[E0412]: cannot find type `DiagnosticBuilder` in this scope
##[error]  --> /checkout/tests/ui-fulldeps/internal-lints/diagnostics.rs:83:48
   |
LL |     fn decorate_lint<'b, >(self, diag: &'b mut DiagnosticBuilder<'a, ()>) {

error[E0412]: cannot find type `DiagnosticBuilder` in this scope
##[error]  --> /checkout/tests/ui-fulldeps/internal-lints/diagnostics.rs:96:46
   |
   |
LL |     fn decorate_lint<'b>(self, diag: &'b mut DiagnosticBuilder<'a, ()>) {

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0412`.
```

</details>

nnethercote commented:
@Nilstrieb I have polished this up and it's ready for review. It's possible that there is a better way to implement parts of `check_expr`. Nonetheless, as written in the PR it's a large improvement over the existing lint implementation, so I think it would be fine to merge in its current state. Please let me know if you are not the right person to review this, and we can find someone more appropriate.

Nilstrieb commented on compiler/rustc_lint/src/internal.rs:
```diff
@@ -364,54 +367,133 @@ declare_lint_pass!(Diagnostics => [UNTRANSLATABLE_DIAGNOSTIC, DIAGNOSTIC_OUTSIDE
+
+                // Is the argument type `impl Into<{Sub,D}iagnosticMessage>`?
+                if let ty::Param(param_ty) = ty.kind()
+                    && let name = param_ty.name.as_str()
```
I think these, uh, checks can just be left out? If someone passes a ty::Param to an Into<DiagnosticMessage> parameter, then the type parameter has to have bounds on these traits.

nnethercote replied:
As per my comment below, the code higher up determines that there is an `impl Into<DiagnosticMessage>` parameter (but isn't sure which ones). Then this code checks all the arguments for one with a type that doesn't require translation. E.g. `String` requires translation, but `DiagnosticMessage` doesn't. Also, an argument of type `impl Into<DiagnosticMessage>` doesn't need translation.

If I remove this entire `if let` (not sure if that's what you mean by "these checks" then I get false positives in methods like this:
```rust
    pub fn buffer_lint(
        &mut self,
        lint: &'static Lint,
        id: NodeId,
        sp: impl Into<MultiSpan>,
        msg: impl Into<DiagnosticMessage>,
    ) {
        self.add_lint(lint, id, sp.into(), msg, BuiltinLintDiagnostics::Normal)
    }
```
where `msg`, which has type `impl Into<DiagnosticMessage>` gets passed along untouched.

Nilstrieb commented on compiler/rustc_lint/src/internal.rs:
```diff
@@ -364,54 +367,133 @@ declare_lint_pass!(Diagnostics => [UNTRANSLATABLE_DIAGNOSTIC, DIAGNOSTIC_OUTSIDE
+        } else if num_impl_into_diagnostic_message_params == 1 {
+            let mut has_diagnostic_message_arg = false;
+            let mut has_impl_into_diagnostic_message_arg = false;
+            for ty in call_tys {
```
How about saving the indices of Into<DiagnosticMessage> params above so you can check just these params? It would also handle multiple parameters properly then.

nnethercote replied:
I was advised to use `predicates_of`, but I'm still puzzled by it. For a call to a `Diag` method like this:
```rust
    pub fn struct_span_err(
        &self,
        span: impl Into<MultiSpan>,
        msg: impl Into<DiagnosticMessage>,
    ) -> Diag<'_>;
```
`predicates_of` returns this:
```
preds: [
    Binder { value: TraitPredicate(<impl Into<MultiSpan> as std::marker::Sized>, polarity:Positive), bound_vars: [] },
    Binder { value: TraitPredicate(<impl Into<DiagnosticMessage> as std::marker::Sized>, polarity:Positive), bound_vars: [] },
    Binder { value: TraitPredicate(<impl Into<MultiSpan> as std::convert::Into<rustc_errors::MultiSpan>>, polarity:Positive), bound_vars: [] },
    Binder { value: TraitPredicate(<impl Into<DiagnosticMessage> as std::convert::Into<rustc_errors::DiagnosticMessage>>, polarity:Positive), bound_vars: [] }
]
```
So the earlier code looks through these and sees the the last element and knows that one of the args is `impl Into<DiagnosticMessage>`. But it doesn't know which one because the indices here don't match the parameter indices.

What I really expected when I started was to instead look up the type signature of `struct_span_err` and go from there. Then I could compare each parameter from the signature against the corresponding argument from the call site. How would I do that? It seems like it might be easier.

nnethercote replied:
Just to reiterate, I'm new to this kind of linting/typechecking code, so things that would be obvious to a more experienced person might be non-obvious to me! I'm just groping my way along.

Nilstrieb replied:
There are two different kinds of "parameter" indices, the type parameter indices (the identity of `ty::Param`) and the function parameter indices (from the `fn_sig()`).
What we want to know for the check is which *function* parameters are  `Into<DiagnosticMessage>`, so we can check those function parameters.
To check whether a function parameter type is `Into<DiagnoticMessage>`, we have two parts. First, the type must be a `ty::Param` (as it's a generic parameter) *and* there must be a trait predicate where *this type* is the self type and the trait is `DiagnosticMessage`.

So I'd propose this pseudocode:
```rust
fn params_that_are_into_diagnostic_message(def_id) -> Vec<bool> {
  let sig = fn_sig(def_id);
  let predicates = predicates_of(def_id);
  let mut is_into_diagnostics_message = vec![false; sig.params().len()]
  for i in 0..sig.params().len() {
    let ty = sig.params()[i];
    if let ty::Param(..) = ty {
      if predicates.any(|p| p.self_ty == ty && p.trait == sym::DiagnosticsMessage) {
        is_into_diagnostics_message[i]  = true;
      }
    }
  }
} 
```

nnethercote replied:
Aha, the `fn_sig` query was the key thing I didn't know about. Once I had that everything fell into place. Thank you!

Nilstrieb commented on compiler/rustc_lint/src/internal.rs:
```diff
@@ -364,54 +367,133 @@ declare_lint_pass!(Diagnostics => [UNTRANSLATABLE_DIAGNOSTIC, DIAGNOSTIC_OUTSIDE
+        }
+
+        // Calls to methods with an `impl Into<{D,Subd}iagnosticMessage>` parameter must be passed
+        // an arg with type `{D,Subd}iagnosticMessage` or `impl Into<{D,Subd}iagnosticMessage`.
```
you didn't close the angle bracket

nnethercote replied:
:+1:

nnethercote commented on compiler/rustc_lint/src/internal.rs:
```diff
@@ -364,54 +367,133 @@ declare_lint_pass!(Diagnostics => [UNTRANSLATABLE_DIAGNOSTIC, DIAGNOSTIC_OUTSIDE
+                    && (
+                        // njn: yuk! how to do this properly?
+                        name.contains("impl Into<DiagnosticMessage>")
+                            || name.contains("impl Into<SubdiagnosticMessage>")
```
Oh, I forgot that I still had this part in here :(

nnethercote commented:
I have updated this to use `fn_sig`, which now means the checking is precise and everything looks much nicer. I think this is ready to go.

davidtwco commented:
@bors r+

bors commented:
:pushpin: Commit 3591e77b3566772023b7e35e96224106243ef214 has been approved by `davidtwco`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=davidtwco 3591e77b3566772023b7e35e96224106243ef214 -->
<!-- homu: {"type":"Approved","sha":"3591e77b3566772023b7e35e96224106243ef214","approver":"davidtwco","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
